kong/tools/queue.lua

```diff
@@ -0,0 +1,408 @@
+    semaphore = semaphore.new(),
+
+    -- `bytes_queued` holds the number of bytes on the queue.  It will be used only if string_capacity is set.
+    bytes_queued = 0,
+    -- `entries` holds the actual queue items.
+    entries = {},
+    -- Pointers into the table that holds the actual queued entries.  `front` points to the oldest, `back` points to
+    -- the newest entry.
+    front = 1,
+    back = 1,
+  }
+  for option, value in pairs(queue_conf) do
+    queue[option] = value
+  end
+
+  queue = setmetatable(queue, Queue_mt)
+
+  kong.timer:named_at(name, 0, function(_, q)
+    while q:count() > 0 do
+      q:log(DEBUG, "processing queue")
+      q:process_once()
+    end
+    q:log(DEBUG, "done processing queue")
+    queues[name] = nil
+  end, queue)
+
+  queues[name] = queue
+
+  queue:log(DEBUG, "queue created")
+
+  return queue
+end
+
+
+-------------------------------------------------------------------------------
+-- Log a message that includes the name of the queue for identification purposes
+-- @param self Queue
+-- @param level: log level
+-- @param formatstring: format string, will get the queue name and ": " prepended
+-- @param ...: formatter arguments
+function Queue:log(level, formatstring, ...)
+  local message = self.name .. ": " .. formatstring
+  if select('#', ...) > 0 then
+    return ngx.log(level, string.format(message, unpack({...})))
+  else
+    return ngx.log(level, message)
+  end
+end
+
+
+function Queue:count()
+  return self.back - self.front
+end
+
+
+-- Delete the frontmost entry from the queue and adjust the current utilization variables.
+function Queue:delete_frontmost_entry()
+  if self.string_capacity then
+    -- If string_capacity is set, reduce the currently queued byte count by the
+    self.bytes_queued = self.bytes_queued - #self.entries[self.front]
+  end
+  self.entries[self.front] = nil
+  self.front = self.front + 1
+  if self.front == self.back then
+    self.front = 1
+    self.back = 1
+  end
+end
+
+-- Process one batch of entries from the queue.  Returns truthy if entries were processed, falsy if there was an
+-- error or no items were on the queue to be processed.
+function Queue:process_once()
+  local ok, err = self.semaphore:wait(0)
+  if not ok then
+    if err ~= "timeout" then
+      -- We can't do anything meaningful to recover here, so we just log the error.
+      self:log(ERR, 'error waiting for semaphore: %s', err)
+    end
+    return
+  end
+  local data_started = now()
+
+  local entry_count = 1
+
+  -- We've got our first entry from the queue.  Collect more entries until max_delay expires or we've collected
+  -- batch_max_size entries to send
+  while entry_count < self.batch_max_size and (now() - data_started) < self.max_delay and not ngx.worker.exiting() do
+    ok, err = self.semaphore:wait(((data_started + self.max_delay) - now()) / 1000)
+    if not ok and err == "timeout" then
+      break
+    elseif ok then
+      entry_count = entry_count + 1
```

Comment:
if there are delays in processing, it means that this function will run slower than producer;
in such a case, why is the code incrementing entry_count rather than taking out max(queued_items, max_batch_size) items at once

Reply:
The `wait()` invocation, according to my reading of the documentation and the behavior that I observed, decrements the semaphore by one.  The documentation does not state that I can supply a negative argument to `post()`, which would be what is required if we wanted to remove a whole batch from the queue with one wait operation.

Once we have this system in place and we have a way to measure performance (the SDETs are working on that), we can start optimizing.  Without a proper measurement setup, I'm not convinced that the additional code complexity is beneficial.

Reply:
Can you not call `count()` and then call `wait()` in a loop appropriately? 
I'm not sure if this code will scale appropriately. I can't explain my hunch well because I lack depth as well. Please talk to someone with deep OpenResty experience and confirm this (a green performance test isn't sufficient to resolve this comment).